### PR TITLE
Use new cache for Integer

### PR DESCRIPTION
**Pull request by debugger22**
See issue https://github.com/sympy/sympy/issues/9292

### CODE REVIEW

**Comment by debugger22** [comment]
Need some help here. Enabling cache for `Integer` has triggered infinite recursion in `Pow` class. @pbrady Could you please see if I am right?

**Comment by asmeurer** [comment]
We should do some benchmarks here once you figure out the test issues. If it significantly slows things down, we shouldn't do it. My main concern with what is there now, aside from code duplication, and it ignoring the settings to the main cache (like `SYMPY_USE_CACHE`) is the memory usage, since the current integer cache grows unbounded.

**Comment by debugger22** [comment]
It works at par with the old cache when fastcache is installed.

**Comment by pbrady** [comment]
@debugger22, the `@cacheit` decorator appears to be applied correctly.  I'm not sure what would cause the infinite recursion but I won't have time to help with debugging this for few days.

**Comment by debugger22** [comment]
I did some debugging and found that this is not being caused by the use of `cacheit` decorator but the removal of special cache for `Integer`.

**Comment by pbrady** [comment]
I wonder if this has something to do with the old cache being initialized
with -1,0,1?

**Comment by debugger22** [comment]
Ideally, it shouldn't matter if I'm not using `return _intcache[ival]` in the constructor. It happens with or without them.

**Comment by pbrady** [comment]
This appears to trigger the infinite recursion
```python
(-I*pi**1)**1
```

**Comment by debugger22** [comment]
Without `fastcache` the special `intcache` is faster but with `fastcache` installed the new cache is marginally better or equivalent.

I used this to test:

```
from sympy import S
ints = []
for i in range(10000):
    ints.append(S(i))

for i in ints:
    i*i + i*i + i + i
```

File: sympy/core/numbers.py [hunk 1]
```diff
@@ -1677,16 +1617,17 @@ def __new__(cls, i):
...
-            # We only work with well-behaved integer types. This converts, for
-            # example, numpy.int32 instances.
-            obj = Expr.__new__(cls)
-            obj.p = ival
-
-            _intcache[ival] = obj
-            return obj
+        # We only work with well-behaved integer types. This converts, for
+        # example, numpy.int32 instances.
+        if ival == 1:
```
Comment (pbrady) [inline sympy/core/numbers.py]: Are these the checks that prevent the infinite recursion?

Comment (debugger22) [inline sympy/core/numbers.py]: Yes.

**Comment by debugger22** [comment]
> Without fastcache the special intcache is faster but with fastcache installed the new cache is marginally better or equivalent.

For a bigger `range` the new cache proves to be much better.

**Comment by debugger22** [comment]
@asmeurer

**Comment by asmeurer** [comment]
If there are no performance regressions I am +1. Maybe we should check this with the new benchmark stuff.

**Comment by debugger22** [comment]
I'm unable to get the asv working with sympy. The cloning process times out. @moorepants Could you please try it once?

**Comment by moorepants** [comment]
Do you have a benchmark you want to run? If you share the code you want to execute we can set it up for asv.

**Comment by debugger22** [comment]
I think what Aaron meant is to run this change in `Integer` with whatever benchmarks we have and compare them.

**Comment by moorepants** [comment]
I see. We don't really have the benchmarks setup to run on pull requests or branches other than master. I'll have to look into that.

But we could just merge and if it causes a performance regression it could be reverted. That's not ideal, but an option. It is essentially what we do already.

**Comment by moorepants** [comment]
I'll run this on the test machine locally to see if there are any changes.

**Comment by debugger22** [comment]
I manually checked out the branch. I think that should give expected results.

```
debugger22@jarvis:~/workspace/sympy_benchmarks$ [move-logic-benchmarks] asv run
· Fetching recent changes.
· Creating environments
· Discovering benchmarks
·· Uninstalling from py2.7-fastcache-mpmath...
·· Building for py2.7-fastcache-mpmath....
·· Installing into py2.7-fastcache-mpmath...
· Running 16 total benchmarks (1 commits * 1 environments * 16 benchmarks)
[  0.00%] · For sympy commit hash 7aa8d4a8:
[  0.00%] ·· Building for py2.7-fastcache-mpmath....
[  0.00%] ·· Benchmarking py2.7-fastcache-mpmath
[  6.25%] ··· Running dsolve.TimeDsolve01.time_dsolve                   873.96ms
[ 12.50%] ··· Running integrate.TimeIntegration01.time_doit             244.01ms
[ 18.75%] ··· Running integrate.TimeIntegration01.time_doit_meijerg      75.58ms
[ 25.00%] ··· Running ...ionOperations.peakmem_jacobian_wrt_functions        34M
[ 31.25%] ··· Running ...ssionOperations.peakmem_jacobian_wrt_symbols        34M
[ 37.50%] ··· Running ...s.TimeLargeExpressionOperations.peakmem_subs        34M
[ 43.75%] ··· Running ...exprs.TimeLargeExpressionOperations.time_cse    44.04ms
[ 50.00%] ··· Running ...essionOperations.time_jacobian_wrt_functions   525.34ms
[ 56.25%] ··· Running ...pressionOperations.time_jacobian_wrt_symbols    10.75ms
[ 62.50%] ··· Running ...perations.time_manual_jacobian_wrt_functions   593.40ms
[ 68.75%] ··· Running ...xprs.TimeLargeExpressionOperations.time_subs   369.66ms
[ 75.00%] ··· Running logic.LogicSuite.time_dpll                           4.04s
[ 81.25%] ··· Running logic.LogicSuite.time_dpll2                       931.28ms
[ 87.50%] ··· Running refine.TimeRefine01.time_refine                      7.09s
[ 93.75%] ··· Running solve.TimeSolve01.time_solve                         4.37s
[100.00%] ··· Running solve.TimeSolve01.time_solve_nocheck                 4.42s
debugger22@jarvis:~/workspace/sympy_benchmarks$ [move-logic-benchmarks] cd sympy/debugger22@jarvis:~/workspace/sympy_benchmarks/sympy$ [rm-autosimp-pow] git checkout Integer-new-cache 
Switched to branch 'Integer-new-cache'
Your branch is up-to-date with 'origin/Integer-new-cache'.
debugger22@jarvis:~/workspace/sympy_benchmarks/sympy$ [Integer-new-cache] cd ..
debugger22@jarvis:~/workspace/sympy_benchmarks$ [move-logic-benchmarks] asv run· Fetching recent changes.
· Creating environments
· Discovering benchmarks
·· Uninstalling from py2.7-fastcache-mpmath.
·· Building for py2.7-fastcache-mpmath
·· Installing into py2.7-fastcache-mpmath...
· Running 16 total benchmarks (1 commits * 1 environments * 16 benchmarks)
[  0.00%] · For sympy commit hash 7aa8d4a8:
[  0.00%] ·· Building for py2.7-fastcache-mpmath....
[  0.00%] ·· Benchmarking py2.7-fastcache-mpmath
[  6.25%] ··· Running dsolve.TimeDsolve01.time_dsolve                        888.97ms
[ 12.50%] ··· Running integrate.TimeIntegration01.time_doit                  240.67ms
[ 18.75%] ··· Running integrate.TimeIntegration01.time_doit_meijerg           73.69ms
[ 25.00%] ··· Running ...pressionOperations.peakmem_jacobian_wrt_functions        34M
[ 31.25%] ··· Running ...ExpressionOperations.peakmem_jacobian_wrt_symbols        34M
[ 37.50%] ··· Running ..._exprs.TimeLargeExpressionOperations.peakmem_subs        34M
[ 43.75%] ··· Running large_exprs.TimeLargeExpressionOperations.time_cse      44.56ms
[ 50.00%] ··· Running ...eExpressionOperations.time_jacobian_wrt_functions   509.43ms
[ 56.25%] ··· Running ...rgeExpressionOperations.time_jacobian_wrt_symbols    10.24ms
[ 62.50%] ··· Running ...sionOperations.time_manual_jacobian_wrt_functions   519.39ms
[ 68.75%] ··· Running large_exprs.TimeLargeExpressionOperations.time_subs    320.28ms
[ 75.00%] ··· Running logic.LogicSuite.time_dpll                                4.04s
[ 81.25%] ··· Running logic.LogicSuite.time_dpll2                            958.04ms
[ 87.50%] ··· Running refine.TimeRefine01.time_refine                           6.85s
[ 93.75%] ··· Running solve.TimeSolve01.time_solve                              4.57s
[100.00%] ··· Running solve.TimeSolve01.time_solve_nocheck                      4.46s
```

**Comment by moorepants** [comment]
I'm not sure that does it. Does your asv config point to the local repo or the github one?

**Comment by debugger22** [comment]
Local repo

**Comment by moorepants** [comment]
But notice that the commit hash is the same for both runs.

**Comment by debugger22** [comment]
Oops. How did that happen. Thanks! Let me try it again.

**Comment by moorepants** [comment]
Maybe try using the explicit commits `commithash1...commithash2` instead of ALL.

**Comment by oscarbenjamin** [comment]
I'm not sure what happened here. I just got this notification although this wan't the PR that I merged... (#15803)

**Comment by asmeurer** [comment]
Apparently GitHub auto marks a PR as merged if you merge another PR based off it. I'll have to think if this can be fixed in the bot. Probably it's best to close a PR if you make a new one off it.

**Comment by asmeurer** [comment]
I guess it could check the merge commit. But this issue only matters for old PRs that were made before the bot. I'll have to check how the API works but it might not be worth worrying about this corner case.

**Comment by oscarbenjamin** [comment]
So maybe it would have been better to close this PR before merging the other?

**Comment by asmeurer** [comment]
Well, I'd say this is a bug in the bot. But presumably GitHub would not have sent a "closed" webhook to the bot if the PR was already closed. Actually it's a good idea to close PRs that are replaced with new ones anyway.

**Pull request merged**